### PR TITLE
Userland: Change class_name to use StringView instead of char const*

### PR DESCRIPTION
**Pull request by ldm5180**
This helps make the overall codebase consistent. `class_name()` in 
`Kernel` is always `StringView`, but not elsewhere.                
                                                                   
Additionally, this results in the `strlen` (which needs to be done 
when printing or other operations) always being computed at        
compile-time.

### CODE REVIEW

**Comment by timschumi** [comment]
Is there a particular reason why we are doing this? Also, we should probably modify the return values to be StringView-literals then.

**Comment by ldm5180** [comment]
> Is there a particular reason why we are doing this? Also, we should probably modify the return values to be StringView-literals then.

Thanks for pointing this out. I didn't realize it would make such a big difference. I will make the changes. https://serenityos.godbolt.org/z/794c8qfK1

**Comment by ldm5180** [comment]
> > Is there a particular reason why we are doing this? Also, we should probably modify the return values to be StringView-literals then.
> 
> Thanks for pointing this out. I didn't realize it would make such a big difference. I will make the changes. https://serenityos.godbolt.org/z/794c8qfK1

Oops. I read this incorrectly. It is what I thought from the beginning. The function without the `sv` literal is more optimize-able. The compiler knows what `strlen` is and can run it at compile-time.

Left side is without the use of `sv`, right side is with `sv`.
https://serenityos.godbolt.org/z/vMvbMYbKd

File: Userland/Libraries/LibCpp/AST.h [hunk 1]
```diff
@@ -144,7 +145,7 @@ class InvalidDeclaration : public Declaration {
 
 public:
     virtual ~InvalidDeclaration() override = default;
-    virtual const char* class_name() const override { return "InvalidDeclaration"; }
+    virtual StringView class_name() const override { return "InvalidDeclaration"; }
```
Comment (tomuta) [inline Userland/Libraries/LibCpp/AST.h:148]: You should probably use the `sv` suffix for string literals that your return as a StringView, saves a runtime `strlen`.

**Review by tomuta — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCpp/AST.h (hunk 1 above)
Comment (ldm5180) [reply Userland/Libraries/LibCpp/AST.h:148]: That is not really true. The optimizer sees through the call to `strlen`. Check out this link to see what actually gets generated: https://serenityos.godbolt.org/z/vMvbMYbKd
Left side is without the use of sv, right side is with sv.

**Comment by trflynn89** [comment]
Interesting, the lack of optimization when adding `sv` seems to be due to the `VERIFY` in the `StringView` constructor. If I comment that out, we get equivalent assembly: https://serenityos.godbolt.org/z/dM1xs5nMK

**Comment by ldm5180** [comment]
> Interesting, the lack of optimization when adding `sv` seems to be due to the `VERIFY` in the `StringView` constructor. If I comment that out, we get equivalent assembly: https://serenityos.godbolt.org/z/dM1xs5nMK

That is not really surprising. Similar thing happens if you make it all `constexpr` so that it is evaluated in a compile-time context: https://serenityos.godbolt.org/z/xTsqc5Gb6

The `Checked::addition_would_overflow` seems like the culprit. It needs to make the jump to the code to handle the case of an overflow. I don't even really know what this means. I guess it is verifying that the address of the pointer + the length doesn't overflow the underlying type of the pointer. This could be avoided by doing something like this where the `sv` is *actually* constructed in a `constexpr` context (https://serenityos.godbolt.org/z/9v3ha74a8). I don't see much benefit though. Might as well just stick with the non-`sv` call which calls the standard c-string constructor, IMHO.

Do you have a preference?

**Comment by trflynn89** [comment]
Since `sv` is so prevalent in the code base already, I think my preference would be to make that case equally optimizable. We can do that with a private constructor that the user-defined literal has friend access to, which can then skip all runtime checks: https://serenityos.godbolt.org/z/15q1fK99s

**Comment by alimpfard** [comment]
@trflynn89 @ldm5180 any reason to avoid `consteval` on `operator""sv`?
that should just make the current code actually work correctly: https://serenityos.godbolt.org/z/7nqaM8jza

**Comment by trflynn89** [comment]
> @trflynn89 @ldm5180 any reason to avoid `consteval` on `operator""sv`? that should just make the current code actually work correctly: https://serenityos.godbolt.org/z/7nqaM8jza

Even better :tada:

**Comment by ldm5180** [comment]
No reason that I know of. I like it!

**Comment by ldm5180** [comment]
This is now rebased onto #13097 and should be ready to merge once it passes.

**Comment by trflynn89** [comment]
Minor thing, each commit message currently contains:
```
Note, the `sv` stringview literal suffix is not used because that
results in code which is not as optimize-able.
```

Which is no longer the case in any of the commits :)

**Comment by trflynn89** [comment]
Also (correct me if I'm wrong) but I think the last two commits aren't atomic? That is, if we keep the LibJS commit, but bisect away the LibWeb commit, the system won't compile, right? If so, those should be one LibJS+LibWeb commit IMO

**Comment by ldm5180** [comment]
> Also (correct me if I'm wrong) but I think the last two commits aren't atomic? That is, if we keep the LibJS commit, but bisect away the LibWeb commit, the system won't compile, right? If so, those should be one LibJS+LibWeb commit IMO

Nice catch! I hadn't noticed the cross-library usage there. Fixed.

**Comment by ldm5180** [comment]
> Minor thing, each commit message currently contains:
> 
> ```
> Note, the `sv` stringview literal suffix is not used because that
> results in code which is not as optimize-able.
> ```
> 
> Which is no longer the case in any of the commits :)

I saw that too. Removed.

**Review by linusg — CHANGES_REQUESTED** [review]
Very nice! The `DevTools` commit needs to be part of the last one actually, that's also a JS Object.

**Comment by ldm5180** [comment]
> Very nice! The `DevTools` commit needs to be part of the last one actually, that's also a JS Object.

Done.

**Comment by linusg** [comment]
Last change then merge: the last commit message is duplicated from the squash :^)

**Comment by ldm5180** [comment]
Ha! Can you tell it is Friday and I am distracted?

Done.

**Pull request merged**